docs/design_proposals/sync-improvements.md

```diff
@@ -0,0 +1,319 @@
+
+#### Suggestion
+These above problems can be made less surprising by a more explicit configuration syntax.
+The `artifact.sync` will be upgraded to a list of *sync rules*.
+
+```yaml
+sync:
+# Existing use-case: Copy all js files under 'src' flat into 'app'
+# This corresponds to the current `'src/**/*.js': app/`
+- from: 'src/**/*.js'
+  to: app/
+  flatten: true  # default to false
+  
+# Existing use-case: Copy all js under 'src' into 'dest', re-creating the directory structure below 'src'  
+# This corresponds to the current `'src/***/*.js': dest/`
+- from: 'src/**/*.js'
+  to: dest/
+  strip: 'src'   # default to ""
+  
+# New use-case: Copy all python files and infer their destination.
+- from: '**/*.py'
+  inferTo: true   # <- tbd, default to false
+```
+
+When determining the destination for a given file, all sync rules will be considered.
+Thus, a single file may be copied to multiple destinations.
+
+Error cases:
+- It should be an error to specify `inferTo=true` together with `to` in the same sync rule.
+  ```yaml
+  - from: '*.js'
+    to: dest/
+    inferTo: true
+  ```
+- It should be an error to specify either of `flatten` or `strip` together with `inferTo`.
+  ```yaml
+  - from: '*.js'
+    inferTo: true
+    flatten: true        # ERR
+  - from: 'src/**/*.js'
+    inferTo: true
+    strip: src/          # ERR
+  ```
+- It should be an error to specify `strip` with a prefix which which does not match the static prefix of the `from` field.
+  ```yaml
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: src/sub        # ERR
+  - from: 'src/**/*.js'
+    to: dest/
+    strip: app/           # ERR
+  ```
+
+Open questions about the configuration:
+
+**\<Is the syntax clear enough?\>**
+Is the syntax clear?
+Is the configuration intuitive to use?
+Is there any ambiguity which needs attention?
+
+Resolution: __Not Yet Resolved__
+
+**\<Does it improve clarity to have the `inferTo` field?\>**
+A different approach could be to fall back to inference, if the `to` field is left blank.
+This would have the advantage that there needs to be no validation about having `to` and `inferTo` in the same rule, which should cause an error.
+
+Resolution: __Not Yet Resolved__
+   
+**\<Error for `strip` and `flatten`\>**
+Given the new sync config, users may try to combine various flags in unintended ways.
+For example, should `strip` and `flatten` in the same rule always be an error?
+My intuition would say, that `strip+flatten` is the same as `flatten`, but this should to be discussed.
+
+Resolution: __Not Yet Resolved__
+
+**\<Do we need an `excludes` rule?\>**
+The current config only suggests additive rules. For some applications it might be easier to also remove matched host paths from the syncable set. For example:
+```yaml
+sync:
+# include all files below src...
+- from: 'src/**/*'
+  to: app/
+# ...but do not consider js files for sync
+- from: 'src/**/*.js'
+  exclude: true
+```
+Also see #1766.
+
+Resolution: __Not Yet Resolved__
```

Comment:
For now, lets go ahead and de-scope it.
With the this Proposal, it will be easy to support this in future.
@balopat WDYT?

Reply:
agreed, we can scope this out for now, create a separate issue to discuss.